eppy/bunch_subclass.py

```diff
@@ -119,73 +143,73 @@ def addfunctions(abunch):
+    }
+    abunch._EpBunch__functions.update(func_dict)
 
     names = [
         'FAN:CONSTANTVOLUME',
         'FAN:VARIABLEVOLUME',
         'FAN:ONOFF',
         'FAN:ZONEEXHAUST',
         'FANPERFORMANCE:NIGHTVENTILATION',
-              ]
-    if key in names:
-        func_dict = {
-            'f_fanpower_bhp': fh.fanpower_bhp,
-            'f_fanpower_watts': fh.fanpower_watts,
-            'f_fan_maxcfm': fh.fan_maxcfm,
-        }
-        abunch.__functions.update(func_dict)
+    ]
+    func_dict = {
+        'f_fanpower_bhp': (fh.fanpower_bhp, names),
+        'f_fanpower_watts': (fh.fanpower_watts, names),
+        'f_fan_maxcfm': (fh.fan_maxcfm, names),
+    }
+    abunch._EpBunch__functions.update(func_dict)
     # =====
     # code for references
-    #-----------------
+    # -----------------
     # add function zonesurfaces
     if key == 'ZONE':
-        func_dict = {'zonesurfaces':fh.zonesurfaces}
-        abunch.__functions.update(func_dict)
+        func_dict = {'zonesurfaces': (fh.zonesurfaces, ['ZONE'])}
+        abunch._EpBunch__functions.update(func_dict)
 
-    #-----------------
+    # -----------------
     # add function subsurfaces
     # going to cheat here a bit
     # check if epbunch has field "Zone_Name"
     # and is in group u'Thermal Zones and Surfaces'
     # then it is likely to be a surface attached to a zone
     fields = abunch.fieldnames
-    try:
-        group = abunch.getfieldidd('key')['group']
-    except KeyError as e:  # some pytests don't have group
-        group = None
-    if group == u'Thermal Zones and Surfaces':
-        if "Zone_Name" in fields:
-            func_dict = {'subsurfaces':fh.subsurfaces}
-            abunch.__functions.update(func_dict)
+    names = ['BuildingSurface:Detailed', 'Wall:Detailed',
+             'RoofCeiling:Detailed', 'Floor:Detailed',
+             'FenestrationSurface:Detailed', 'Shading:Site:Detailed',
+             'Shading:Building:Detailed', 'Shading:Zone:Detailed']
+    func_dict = {'subsurfaces': (fh.subsurfaces, names)}
+    abunch._EpBunch__functions.update(func_dict)
 
     return abunch
 
+
 class EpBunch(Bunch):
     """
     Fields, values, and descriptions of fields in an EnergyPlus IDF object
     stored in a `bunch` which is a `dict` extended to allow access to dict
     fields as attributes as well as by keys.
 
     """
+    __functions = {}
+
     def __init__(self, obj, objls, objidd, *args, **kwargs):
         super(EpBunch, self).__init__(*args, **kwargs)
         self.obj = obj  # field names
         self.objls = objls  # field values
         self.objidd = objidd  # field metadata (minimum, maximum, type, etc.)
         self.theidf = None  # pointer to the idf this epbunch belongs to
-                              # This is None if there is no idf - a standalone epbunch
-                              # This will be set by Idf_MSequence
-        self['__functions'] = {}  # initialize the functions
```

Comment:
Moving the `__functions` attribute to the class allows to register functions before class instantiation. Without this change, one would have to create an EpBunch instance and then add the method to it.